### ISSUE TITLE
Request made immediately after calling stop on the service worker doesn't always passthrough properly

### ISSUE THREAD

**Issue by spocke** [kettanaito/msw#2323]
### Prerequisites

- [X] I confirm my issue is not in the [opened issues](https://github.com/mswjs/msw/issues)
- [X] I confirm the [Frequently Asked Questions](https://mswjs.io/docs/faq) didn't contain the answer to my issue

### Environment check

- [X] I'm using the [latest](https://github.com/mswjs/msw/releases/latest) `msw` version
- [X] I'm using Node.js version 18 or higher

### Browsers

Chromium (Chrome, Brave, etc.), Safari

### Reproduction repository

https://github.com/spocke/mswjs-stop-bug

### Reproduction steps

1. Clone the attached repo
1. Install the npm dependencies using `npm i`
2. Start the server using `npm start`
3. Open the Safari or Chrome and go to go to `http://localhost:3000`
4. Re-load the page N times might take a few tries
5. Observe that sometimes the second image is not served (network tabs shows it in a pending state)
6. Observe that sometimes the second image is served though the service worker but not properly served it just spins and spins forever in the network tab

### Current behavior

Resources are still served though an active client rather than passed though the service worker even if you call `worker.stop()` sometimes that even fails to serve the request since likely because of the handlers being removed for it on the parent page side.
It's likely flakey like this because the post message from the browser to the service worker takes a while to process so it can't be a fire and forget it needs to be async and send a ACK back when it has properly intercepted the message to stop intercepting requests.

### Expected behavior

No requests should be processed though an active client on the service worker once you call stop on it. The stop function probably needs to be async to know when it's safe to do new requests and expect them to be passed though by https://github.com/mswjs/msw/blob/ca6cb7e8e95ad76c6fe6c44d7f4b80d36bc53478/src/mockServiceWorker.js#L109.

**Comment by kettanaito** [comment]
Hi, @spocke. Thanks for reporting this. 

## What `worker.stop()` does

Note that `worker.stop()` doesn't mean unregistering the worker. I've tried to highlight that in the docs as well:

<img width="882" alt="Screenshot 2024-10-21 at 17 31 11" src="https://github.com/user-attachments/assets/7cec590e-534a-467a-9e9a-95241ca43e70">

Your site's resources will be processed by the worker because it remains registered and active. The `worker.stop()` call tells the worker to not include your request handlers during that processing. It's also _client-scoped_, and you can have different tabs of the same app having the interception enabled and disabled at the same time. 

Once you called `worker.stop()`, the current client's ID is removed from the internal list of clients in the worker. Any requests originating from this client are ignored (i.e. processed as-is). 

## But why not unregister the worker?

Registering the worker is a costly action. That's why MSW automatically unregisters the worker once you close the last active controlled client (i.e. the last tab of your app). Due to how Service Workers operate, there is no difference between having no worker and having a worker that bypasses all the network. 

## But what if I need to unregister the worker?

Then you can do so via:

```js
await navigator.serviceWorker.controller.getRegistration((registration) => {
  return registration.unregister()
})
```

I see little reason to do that, however. Debugging may be the only reason I can think of.

**Comment by kettanaito** [comment]
Given the context above, can you rephrase the actual/expected behaviors so it's more clear?

**Comment by spocke** [comment]
Sorry for any confusion here but I'm not talking about unregister the service worker. From my understanding it remains registered but goes into an inactive state so that all requests that gets served though the worker just passes though. However the issue here is that once you post a message to the worker the stop processing requests that code assumes that the postMessage to the worker is instant that is not the case it takes a few milliseconds then it removes any local handlers for it. So it's in this limbo state for a while.
So the workaround I have is to just issue a integrity check to the service worked since that is a request type of action so I can know that the service worker has received and processed the request I know that it also processed the previous in flight stop fire and forget call. The other workaround would be to call stop then wait for N time until the message has been processed by the worker but how long do you wait it's up to the browser to schedule post messages so that might take random time to finish so not ideal.

**Comment by spocke** [comment]
So from my understanding it does this:
https://github.com/mswjs/msw/blob/020161ef720840efd8742fe9e49326abf54f0114/src/browser/setupWorker/setupWorker.ts#L185
So unregisters all local handlers and postMessage a `MOCK_DEACTIVATE` https://github.com/mswjs/msw/blob/ca6cb7e8e95ad76c6fe6c44d7f4b80d36bc53478/src/mockServiceWorker.js#L70
But since it's sync this will now be in a limbo state where the client has not been removed by the service worker yet but all the handlers have been removed on the parent page so there is nothing that can serve the requests anymore until the message has been processed by the worker and the client is properly removed then it goes into the pass though state in https://github.com/mswjs/msw/blob/ca6cb7e8e95ad76c6fe6c44d7f4b80d36bc53478/src/mockServiceWorker.js#L109 where it has no clients.

**Comment by kettanaito** [comment]
Thanks for a more detailed explanation. I think I got it now. 

So the issue is the window of an unexpected behavior between calling `worker.stop()` and the client ID actually being removed from the internal worker state. 

Making `worker.stop()` return a Promise will be a breaking change. Instead, I suggest checking `worker.context.isMockingEnabled` in the request listener on the client:

https://github.com/mswjs/msw/blob/020161ef720840efd8742fe9e49326abf54f0114/src/browser/setupWorker/start/createRequestListener.ts#L17

Basically, if we arrive at that problematic window, and the worker sends a request to the client while removing its ID from the set, the client can ignore that request event if `isMockingEnabled` is `false`. 

It would be nice to design a reliable test case for this.

**Comment by kettanaito** [comment]
The thing with that window is that even if a request happens within it, the client doesn't remove any handlers. It will still process that request. The only layer that is affected by `worker.stop()` is this check:

https://github.com/mswjs/msw/blob/ca6cb7e8e95ad76c6fe6c44d7f4b80d36bc53478/src/mockServiceWorker.js#L109

But this check happens in response to any request immediately. So upon an intercepted request, there's either some client to handle it, or not. If there's a client, the worker will message it and get the result back (no matter if `worker.stop()` has somehow happened in parallel). If there isn't any clients, the request is ignored. 

Perhaps there's something else to this issue I'm missing.

**Comment by spocke** [comment]
I'm pretty new to this project but the bug is very easy to reproduce with the repo I made. With that it clear that requests are not processed properly by the service worker after calling stop on it.

I see no error messages when it's not processing the request it's just spinning with `pending` so it's like the client is still connected but it's waiting for a response from the parent page to process the request and that never happens so it just hangs.

So here is a screenshot of a failed session the second request just spins forever:
![image](https://github.com/user-attachments/assets/f76f9c60-0a68-4707-8ead-cce2a97b6916)

It's pretty basic it loads two images one while the mocking is active and one after the `stop` call has been made. The second request hangs sometimes especially on Safari but happens on Chrome as well.

**Comment by kettanaito** [comment]
My hope is that Chrome dispatches the [`controllerchange`](https://developer.mozilla.org/en-US/docs/Web/API/ServiceWorkerContainer/controllerchange_event) event when it re-registers the idle worker. That sounds like a fitting event to dispatch—the client has acquired a new worker. 

Unfortunately, no matter the idle timeout (5m, 10m), this is not reproducible in Playwright. I suspect this behavior is not a part of Chromium but Chrome. Testing it manually in Chrome then.

**Comment by kettanaito** [comment]
@spocke, can you trace what's going on if you put a `debugger` inside the `fetch` listener in `mockServiceWorker.js`? Please share how far you get and whether something will seem suspicious during the second, problematic image resolution.

**Comment by spocke** [comment]
Ok so I added logging in the `message` event handler function when the `fetch` event handler is called and before and after the `getResponse` call in handle request I also logging when the parent page/client is sending the `MOCK_DEACTIVATE` however only showing that for the Chrome logs since they share the same console output it easier to spot the difference.

So from that I can see that the failing ones have the `fetch` call in flight while the `MOCK_DEACTIVATE` call is made to the service worker but not yet received. So I think my assumption about the race condition was correct, the stop needs to be async and wait for the message to be fully processed by the worker until one can assume that it's properly switched off or some other workaround for unhandled in flight requests needs to be added.

Other observations Safari logs `DataCloneError` not sure what that is about. Also I noticed that the `worker.context.requests` map is not fully cleared on Safari for some reason so might be a memory leak there?

Failed Safari:
![image](https://github.com/user-attachments/assets/2c1a7e83-dff4-4f11-add0-7ba66a9178ec)

Passed SafarI:
![image](https://github.com/user-attachments/assets/435cf9b8-06fe-44a6-8911-4f837e4cfb4b)

Passed Chrome:
![image](https://github.com/user-attachments/assets/35106107-1045-4de4-b096-9b815be37463)

Failed Chrome:
![image](https://github.com/user-attachments/assets/0e06b94d-d628-4cde-9083-722e040b08bc)

**Comment by spocke** [comment]
@kettanaito Do you need anything more from me. I would be good to get this one fixed so I don't need to have ugly workarounds in the integration of this library.

**Comment by TheSpyder** [comment]
> Making `worker.stop()` return a Promise will be a breaking change

Only if a lint rule like `no-floating-promises` is active. TypeScript by default shouldn't break if a function called as void returns a value.

Without knowing much about the internals... I wonder if instead of `stop()` assuming it can disconnect instantly, could it instead swap to some sort of "inactive" state that will bypass new requests, but in-flight requests are still able to resolve?

**Comment by TheSpyder** [comment]
I think I've come up with a more reproducible test case based on @spocke's repo, by creating the second image tag _before_ `stop` is called but introducing a 1 second mock delay to the `passthrough` response. This more clearly shows how the bug is a race condition within the `worker` state management.

https://github.com/TheSpyder/mswjs-stop-bug

Interestingly, it passes every time - including the passthrough delay - if the pause before calling `stop` is half a second. Perhaps the race condition is not quite where I thought it was.

**Comment by kettanaito** [comment]
> could it instead swap to some sort of "inactive" state that will bypass new requests, but in-flight requests are still able to resolve?

I believe this is how it should work right now, but that's apparently not the case.

`MOCK_DEACTIVATE` in the worker only does one thing: removes the client that sent that message from the map of active clients. That map is only relevant in one place: when _receiving_ new `fetch` events. If a fetch event is already pending, removing the client that sent it from the active clients map will have no effect on that pending request's resolution. 

## Scenarios

Here's the intended behavior of different scenarios.

### In-flight + stop

If a fetch request is in-flight and the client called `worker.stop()`:

1. The in-flight request resolves according to your handlers;
2. Any subsequent requests from the same client are unmocked.

### Stop + new fetch

If the client calls `worker.stop()` and then issues a new fetch request immediately after:

1. The new request must be unmocked, regardless if the worker confirms deactivation (it doesn't since there's nothing to confirm).

> [!IMPORTANT]
> I understand this scenario is the one that's faulty right now. Removing the client from the active clients map is async due to the client-worker messaging so by the time a new request fires, `activeClients` might still faultily contain the client ID that called `worker.stop()`.

**Comment by kettanaito** [comment]
I've come to realize a few fundamental issues with stopping the setupWorker instance. 

First, it was always meant as a client-side state. Whether the setup worker affects the network is something only the client controls, never the worker. For the worker, it is irrelevant. We should remove the `MOCK_DEACTIVATE` event altogether, it's not needed.

Removing it and moving the state client-side where it belongs solves the second issue: stop, then fetch. Calling `worker.stop()` sets `context.isMockingEnabled = false` immediately, so any requests happening past that point will always be passthrough.

The first issue, however, still remains as the race condition between a request firing and the `REQUEST` message arriving from the worker is still possible. Since the client never knows when requests happen, I'm considering adding a timestamp to the `REQUEST` event payload so we could compare the time when the request was caught by the worker and when `worker.stop()` was called. 

```ts
.on('REQUEST', (event) => {
  if (event.data.issuedAt >= this.context.stoppedAt) {
    return event.postMessage('PASSTHROUGH')
  }
})
```

## Another issue

A semi-related issue here is that it must be forbidden to have multiple `setupWorker()` calls in the same browser context. My fix is based on keeping the `REQUEST` listener from the worker. The worker keeps affecting the traffic of the clients even if they were stopped. The only thing that stop actually does is _ignores handler lookup_ (i.e. always passthrough). 

Which means that in case of multiple `setupWorker()` insatnces, you end up with multiple `REQUEST` listeners attached to the worker controller, and that yields non-deterministic network behavior. 

This isn't strictly related to this issue but something I discovered while looking into it. We should forbid multiple `setupWorker()` calls.

**Comment by kettanaito** [comment]
What I ended up doing is timestamping the `fetch` event in the worker (https://github.com/mswjs/msw/pull/2578/commits/0617b266719c834649900244326f3857d6ec91b0) and comparing it to the timestamp of when `worker.stop()` was called. This isn't ideal as it's prone to resulting in false-negative comparisons if it takes the browser some time to dispatch the `fetch` event on the worker (nothing we can do about this time). Considering MSW's limitations (requests aren't tracked client-side), this is the best we can do. This fix has both in-flight scenarios passing (before and after calling `worker.stop()`).

**Comment by kettanaito** [comment]
## Released: v2.11.2 🎉

This has been released in v2.11.2!

- 📄 [**Release notes**](https://github.com/mswjs/msw/releases/tag/v2.11.2)
- 📦 [npm package](https://www.npmjs.com/package/msw/v/2.11.2)

Make sure to always update to the latest version (`npm i msw@latest`) to get the newest features and bug fixes.

---

_Predictable release automation by [@ossjs/release](https://github.com/ossjs/release)_.
